### PR TITLE
fix: Remove circular dependencies

### LINKED ISSUE #1448: [React Native]: Require cycle warning
### New Issue Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/Parse-SDK-JS/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the latest versions of [Parse JS SDK](https://github.com/parse-community/Parse-SDK-JS/releases).

### Issue Description
Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.

### Steps to reproduce
Run with React Native project

### Expected Outcome
Not to have require cycle, resolve the warning

###  Environment
```
System:
    OS: macOS 12.2
    CPU: (8) arm64 Apple M1
    Memory: 84.22 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.17.3 - ~/.nvm/versions/node/v14.17.3/bin/node
    Yarn: 1.22.17 - ~/.yarn/bin/yarn
    npm: 6.14.13 - ~/.nvm/versions/node/v14.17.3/bin/npm
    Watchman: 2022.01.31.00 - /opt/homebrew/bin/watchman
  Managers:
    CocoaPods: 1.11.2 - /opt/homebrew/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 21.2, iOS 15.2, macOS 12.1, tvOS 15.2, watchOS 8.3
    Android SDK:
      API Levels: 28, 29, 30, 31
      Build Tools: 27.0.3, 29.0.2, 30.0.0, 30.0.2, 31.0.0
      System Images: android-25 | Android Wear ARM EABI v7a, android-30 | Google Play ARM 64 v8a
      Android NDK: Not Found
  IDEs:
    Android Studio: 2020.3 AI-203.7717.56.2031.7621141
    Xcode: 13.2.1/13C100 - /usr/bin/xcodebuild
  Languages:
    Java: 1.8.0_292 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.2 => 17.0.2 
    react-native: 0.66.4 => 0.66.4 
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```

Client
- Parse JS SDK version: 3.4.0

### Logs

```
WARN  Require cycle: node_modules/parse/lib/react-native/ParseACL.js -> node_modules/parse/lib/react-native/ParseRole.js -> node_modules/parse/lib/react-native/ParseACL.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/parse/lib/react-native/ParseObject.js -> node_modules/parse/lib/react-native/canBeSerialized.js -> node_modules/parse/lib/react-native/ParseObject.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/parse/lib/react-native/ParseObject.js -> node_modules/parse/lib/react-native/canBeSerialized.js -> node_modules/parse/lib/react-native/ParseRelation.js -> node_modules/parse/lib/react-native/ParseObject.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/parse/lib/react-native/ParseACL.js -> node_modules/parse/lib/react-native/ParseRole.js -> node_modules/parse/lib/react-native/ParseObject.js -> node_modules/parse/lib/react-native/canBeSerialized.js -> node_modules/parse/lib/react-native/ParseRelation.js -> node_modules/parse/lib/react-native/ParseQuery.js -> node_modules/parse/lib/react-native/encode.js -> node_modules/parse/lib/react-native/ParseACL.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/parse/lib/react-native/ParseObject.js -> node_modules/parse/lib/react-native/canBeSerialized.js -> node_modules/parse/lib/react-native/ParseRelation.js -> node_modules/parse/lib/react-native/ParseQuery.js -> node_modules/parse/lib/react-native/encode.js -> node_modules/parse/lib/react-native/ParseObject.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/parse/lib/react-native/ParseRelation.js -> node_modules/parse/lib/react-native/ParseQuery.js -> node_modules/parse/lib/react-native/encode.js -> node_modules/parse/lib/react-native/ParseRelation.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/parse/lib/react-native/ParseObject.js -> node_modules/parse/lib/react-native/canBeSerialized.js -> node_modules/parse/lib/react-native/ParseRelation.js -> node_modules/parse/lib/react-native/ParseQuery.js -> node_modules/parse/lib/react-native/ParseObject.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/parse/lib/react-native/decode.js -> node_modules/parse/lib/react-native/ParseACL.js -> node_modules/parse/lib/react-native/ParseRole.js -> node_modules/parse/lib/react-native/ParseObject.js -> node_modules/parse/lib/react-native/decode.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/parse/lib/react-native/ParseObject.js -> node_modules/parse/lib/react-native/EventuallyQueue.js -> node_modules/parse/lib/react-native/ParseObject.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/parse/lib/react-native/ParseACL.js -> node_modules/parse/lib/react-native/ParseRole.js -> node_modules/parse/lib/react-native/ParseObject.js -> node_modules/parse/lib/react-native/ParseACL.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/parse/lib/react-native/ParseObject.js -> node_modules/parse/lib/react-native/SingleInstanceStateController.js -> node_modules/parse/lib/react-native/ObjectStateMutations.js -> node_modules/parse/lib/react-native/ParseObject.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/parse/lib/react-native/ParseObject.js -> node_modules/parse/lib/react-native/unique.js -> node_modules/parse/lib/react-native/arrayContainsObject.js -> node_modules/parse/lib/react-native/ParseObject.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/parse/lib/react-native/ParseObject.js -> node_modules/parse/lib/react-native/unique.js -> node_modules/parse/lib/react-native/ParseObject.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/parse/lib/react-native/ParseObject.js -> node_modules/parse/lib/react-native/unsavedChildren.js -> node_modules/parse/lib/react-native/ParseObject.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/parse/lib/react-native/ParseUser.js -> node_modules/parse/lib/react-native/ParseSession.js -> node_modules/parse/lib/react-native/ParseUser.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
```

### PR DESCRIPTION
**Pull request by dplewis**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/Parse-SDK-JS/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/Parse-SDK-JS/blob/alpha/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/Parse-SDK-JS/issues?q=is%3Aissue).

## Issue

The SDK has circular dependencies, require cycles are allowed, but can result in uninitialized values. This prevents the SDK from moving to ES6 for tree-shaking and full Typescript support.

```
1) ParseACL.js > ParseRole.js
2) ParseObject.js > ParseACL.js > ParseRole.js
3) ParseUser.js > ParseObject.js > ParseACL.js
4) ParseObject.js > ParseOp.js
5) ParseObject.js > ParseOp.js > ParseRelation.js
6) ParseOp.js > ParseRelation.js
7) ParseObject.js > ParseOp.js > ParseRelation.js > ParseQuery.js > OfflineQuery.js > decode.js
8) ParseOp.js > ParseRelation.js > ParseQuery.js > OfflineQuery.js > decode.js
9) ParseRelation.js > ParseQuery.js > OfflineQuery.js > decode.js
10) ParseObject.js > ParseOp.js > ParseRelation.js > ParseQuery.js > OfflineQuery.js > equals.js
11) ParseObject.js > ParseOp.js > ParseRelation.js > ParseQuery.js
12) ParseOp.js > ParseRelation.js > ParseQuery.js > encode.js
13) ParseRelation.js > ParseQuery.js > encode.js
14) ParseObject.js > ParseOp.js > arrayContainsObject.js
15) ParseObject.js > SingleInstanceStateController.js
16) uuid.js
```

Closes: https://github.com/parse-community/Parse-SDK-JS/issues/1448

## Approach

* Using CoreManager to centralize modules for refactoring
* Add Op, User, Query, Object and Role to CoreManager
* Convert `uuid.js` to typescript
* Remove localStorage mock as jest automatically mocks it
* Add new circular dependency check to CI

Shoutout to @swittk for coming up with the idea and testing it out.

## Tasks

- [x] Add tests
- [x] Add changes to documentation (guides, repository pages, code comments)

### CODE REVIEW

**Comment by mtrezza** [comment]
Any idea why our "check circular" job didn't detect these?

**Comment by dplewis** [comment]
There is a check circular job on Parse Server but not in the SDK.

File: .madgerc [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+	"detectiveOptions": {
+		"ts": {
+			"skipTypeImports": true
+		},
+    "es6": {
+			"skipTypeImports": true
+		}
+	}
+}
```
Comment (mtrezza) [inline .madgerc]:
```suggestion
{
  "detectiveOptions": {
    "ts": {
      "skipTypeImports": true
    },
    "es6": {
      "skipTypeImports": true
    }
  }
}
```

**Review by mtrezza — APPROVED** [review]

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.1.0-alpha.10](https://github.com/parse-community/Parse-SDK-JS/releases/tag/5.1.0-alpha.10)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.1.0-beta.1](https://github.com/parse-community/Parse-SDK-JS/releases/tag/5.1.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.1.0](https://github.com/parse-community/Parse-SDK-JS/releases/tag/5.1.0)

**Pull request merged**